src/main/java/org/codehaus/groovy/ast/tools/GenericsUtils.java

```diff
@@ -950,7 +956,7 @@ public static boolean hasUnresolvedGenerics(final ClassNode type) {
     public static Tuple2<ClassNode[], ClassNode> parameterizeSAM(final ClassNode samType) {
         MethodNode abstractMethod = ClassHelper.findSAM(samType);
 
-        Map<GenericsType, GenericsType> generics = makeDeclaringAndActualGenericsTypeMapOfExactType(abstractMethod.getDeclaringClass(), samType);
+        Map<GenericsType, GenericsType> generics = makeDeclaringAndActualGenericsTypeMap(abstractMethod.getDeclaringClass(), samType);
```

Comment:
We have been moving away from `makeDeclaringAndActualGenericsTypeMap` towards `makeDeclaringAndActualGenericsTypeMapOfExactType`.  Can you explain this change?

Reply:
Is this the stuff that sits under parametwrizeSAM?  That was the key method for fixing lambda param types.

Reply:
I'm sorry, this was a mistake. 
I will revert this change.

Reply:
I'm sorry, this change was not intended to be commited, it was just for testing. 
I will revert it.

Reply:
> Is this the stuff that sits under parametwrizeSAM? That was the key method for fixing lambda param types.

parameterizeSAM uses this makeDeclaringAndActualGenericsTypeMapOfExactType method to map declaring generics to actual generics. I thought that I needed to adjust this mapping so for unresolved generics it will output Object, but I'm still unsure if that's the right solution (as you mentioned in the other comment).